cmd/thanos/compact.go

```diff
@@ -325,14 +356,25 @@ func runCompact(
 			level.Warn(logger).Log("msg", "downsampling was explicitly disabled")
 		}
 
-		if err := compact.ApplyRetentionPolicyByResolution(ctx, logger, bkt, metaFetcher, retentionByResolution); err != nil {
+		if err := compact.ApplyRetentionPolicyByResolution(ctx, logger, bkt, metaFetcher, retentionByResolution, blocksMarkedForDeletion); err != nil {
 			return errors.Wrap(err, fmt.Sprintf("retention failed"))
 		}
 
-		compact.BestEffortCleanAbortedPartialUploads(ctx, logger, metaFetcher, bkt, partialUploadDeleteAttempts)
+		compact.BestEffortCleanAbortedPartialUploads(ctx, logger, metaFetcher, bkt, partialUploadDeleteAttempts, blocksMarkedForDeletion)
 		return nil
 	}
 
+	g.Add(func() error {
+		return runutil.Repeat(deleteDelay, ctx.Done(), func() error {
+			if err := blocksCleaner.DeleteMarkedBlocks(ctx); err != nil {
```

Comment:
We have a race here as @GiedriusS found (:

Reply:
Maybe we should simplify and avoid concurrency after all. (:

Reply:
I am now doing this as part of the `compactorMainFn` itself, this will make sure that the race conditions are not there.

Reply:
Please create a thread if the approach is not the correct approach. Thanks